src/control/cmd/dmg/storage.go

```diff
@@ -19,13 +19,14 @@ import (
 
 // storageCmd is the struct representing the top-level storage subcommand.
 type storageCmd struct {
-	Scan       storageScanCmd     `command:"scan" description:"Scan SCM and NVMe storage attached to remote servers."`
-	Format     storageFormatCmd   `command:"format" description:"Format SCM and NVMe storage attached to remote servers."`
-	Query      storageQueryCmd    `command:"query" description:"Query storage commands, including raw NVMe SSD device health stats and internal blobstore health info."`
-	NvmeRebind nvmeRebindCmd      `command:"nvme-rebind" description:"Detach NVMe SSD from kernel driver and rebind to userspace driver for use with DAOS."`
-	Set        setFaultyCmd       `command:"set" description:"Manually set the device state."`
-	Replace    storageReplaceCmd  `command:"replace" description:"Replace a storage device that has been hot-removed with a new device."`
-	Identify   storageIdentifyCmd `command:"identify" description:"Blink the status LED on a given VMD device for visual SSD identification."`
+	Scan          storageScanCmd     `command:"scan" description:"Scan SCM and NVMe storage attached to remote servers."`
+	Format        storageFormatCmd   `command:"format" description:"Format SCM and NVMe storage attached to remote servers."`
+	Query         storageQueryCmd    `command:"query" description:"Query storage commands, including raw NVMe SSD device health stats and internal blobstore health info."`
+	NvmeRebind    nvmeRebindCmd      `command:"nvme-rebind" description:"Detach NVMe SSD from kernel driver and rebind to userspace driver for use with DAOS."`
+	NvmeAddDevice nvmeAddDeviceCmd   `command:"nvme-add-device" description:"Add a hot-inserted NVMe SSD to a specific engine configuration to enable the new device to be used."`
+	Set           setFaultyCmd       `command:"set" description:"Manually set the device state."`
+	Replace       storageReplaceCmd  `command:"replace" description:"Replace a storage device that has been hot-removed with a new device."`
```

Comment:
Could you please advise why "replace" isn't enough and we need to expose yet another command?

Reply:
@NiuYawei probably best to answer this as he created the requirements. From my own understanding this command is to facilitate the situation where we have added a device which initially cannot be communicated with from the engine because it is not bound to userspace driver. The control-plane does not yet have capability to react to hotplug events (that is in development) so firstly admin calls "dmg storage nvme-rebind" to bind to userspace and the second command "dmg storage nvme-add-device" to update nvme.conf on SCM for a specific engine so the new device is available on restart.

If we wanted to contain all the necessary functionality within Replace it would get pretty messy IMO because of needing to specify so many parameters (engine index, tier index, PCI address, SMD old and new UUIDs), additionally BIO will not be able to generate SMD UUIDs until device is rebound to userspace after being plugged so currently we need distinct commands.

There is detail in the JIRA ticket and it's parent.

Please note that these commands are eventually going to be reserved for hotplug override for situations where the automatic features don't work, the internal functions will be triggered automatically when we have the capability to respond to PCI (or udev) events in the control plane. This work is captured in the following tickets and effort is relatively significant partly due to the recently added requirement of removing the NVMe caching in the control plane.

Reply:
https://daosio.atlassian.net/browse/DAOS-9921 https://daosio.atlassian.net/browse/DAOS-9853

Reply:
I wonder, would it make sense to combine the rebind command and the add-nvme command into a single command? Are there circumstances where we'd want to rebind a device without immediately adding it to DAOS?

Reply:
There is no practical reason why these commands cannot be merged, the ticketed requirements were for 2 separate commands and I assume this was to coincide with requirements for the hotplug demo. I could merge both into "NvmeAddDevice" now or do that later in the 2.2-2.4 development timeframe as the requirement has been pushed out to 2.4.

Reply:
I think it's better to make them into one command, but I agree with you that we can leave them as separate commands for experimentation.